File: tests/quitus-pdf.php
<?php
/**
 * PDF de contrôle sans accès à WordPress ni à la base : données fictives uniquement.
 * Usage : /opt/lampp/bin/php tests/quitus-pdf.php [répertoire de sortie]
 * Puis : python3 tests/quitus-pdf.py [même répertoire]
 */
if ( PHP_SAPI !== 'cli' ) {
	exit;
}
define( 'ABSPATH', dirname( __DIR__ ) . '/' );
define( 'UEB_INSC_DIR', dirname( __DIR__ ) );
define( 'K_PATH_CACHE', sys_get_temp_dir() . '/' );
foreach ( array( 'config', 'nombres', 'quitus', 'quitus-pdf' ) as $module ) {
	require UEB_INSC_DIR . '/inc/' . $module . '.php';
}
require UEB_INSC_DIR . '/lib/tcpdf/tcpdf.php';

function pdf_de_controle() {
	$pdf = new TCPDF( 'P', 'mm', 'A4', true, 'UTF-8', false );
	$pdf->setFontSubsetting( false );
	$pdf->setPrintHeader( false );
	$pdf->setPrintFooter( false );
	$pdf->SetMargins( 0, 0, 0 );
	$pdf->SetAutoPageBreak( false );
	$pdf->setCellPaddings( 0, 0, 0, 0 );
	$pdf->setCellHeightRatio( 1.12 );
	return $pdf;
}

$sortie = $argv[1] ?? sys_get_temp_dir() . '/ueb-quitus-qr-review';
if ( ! is_dir( $sortie ) ) {
	mkdir( $sortie, 0700, true );
}
$recueil = pdf_de_controle();
$manifest = array();
foreach ( ueb_etablissements() as $sigle => $etab ) {
	foreach ( array( 'droits', 'medicaux' ) as $type ) {
		$medical = 'medicaux' === $type;
		$q = (object) array(
			'numero' => 'TEST-' . $sigle . ( $medical ? '-M' : '' ) . '-2627-0001',
			'etablissement' => $sigle, 'annee_academique' => '2026-2027', 'type' => $type,
			'type_identifiant' => $medical ? 'matricule' : 'dossier',
			'identifiant' => $medical ? '24TEST01' . $sigle : 'UEB-2026-999999',
			'nom' => 'ÉTUDIANT TEST', 'prenom' => 'Marie-Anne Élodie',
			'date_naissance' => '2002-04-12', 'lieu_naissance' => 'Ebolowa',
			'sexe' => 'F', 'nationalite' => 'Camerounaise',
			'departement' => 'Biotechnologie et pharmacognosie', 'parcours' => 'M1',
			'montant' => $medical ? 5000 : 50000, 'tranche' => $medical ? 0 : 3,
			'email' => 'non-imprime@example.com', 'nom_urgence' => 'Contact privé',
		);
		$pdf = pdf_de_controle();
		ueb_pdf_page_quitus( $pdf, $q );
		ueb_pdf_page_quitus( $recueil, $q );
		$fichier = $sigle . '-' . $type . '.pdf';
		$pdf->Output( $sortie . '/' . $fichier, 'F' );
		$manifest[] = array( 'fichier' => $fichier, 'couleur' => $etab['couleur'], 'quitus' => $q, 'qr' => ueb_pdf_contenu_qr( $q ) );
	}
}
$recueil->Output( $sortie . '/quitus-couleurs-qr-18-pages.pdf', 'F' );

// Vérifier aussi une longue identité et une deuxième tranche avec accents/apostrophes.
$q = clone $q;
$q->type = 'droits';
$q->numero = 'TEST-ENSTMO-2627-0002';
$q->nom = 'ÉTUDIANT TEST NOM COMPOSÉ NGONO ESSOMBA';
$q->prenom = 'Marie-Anne Élodie Grâce Charlotte';
$q->lieu_naissance = 'Nko’ovos — Ebolowa';
$q->departement = 'Sciences et techniques maritimes, océanographie et gestion des ressources halieutiques';
$q->tranche = 2;
$q->montant = 25000;
$pdf = pdf_de_controle();
ueb_pdf_page_quitus( $pdf, $q );
$pdf->Output( $sortie . '/identite-longue.pdf', 'F' );
$manifest[] = array( 'fichier' => 'identite-longue.pdf', 'couleur' => $etab['couleur'], 'quitus' => $q, 'qr' => ueb_pdf_contenu_qr( $q ) );
file_put_contents( $sortie . '/manifest.json', json_encode( $manifest, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE ) );
echo "PDF de contrôle générés : $sortie\n";

File: tests/quitus-pdf.py
"""Contrôle des PDF réels : couleurs d'en-tête et lecture des quatre QR à 300 dpi.

Prérequis : pypdf, pypdfium2, Pillow et zxing-cpp.
Exécuter auparavant tests/quitus-pdf.php. Aucune connexion à la base.
"""
import json
import sys
from collections import Counter
from pathlib import Path

import pypdfium2 as pdfium
import zxingcpp
from pypdf import PdfReader

root = Path(sys.argv[1] if len(sys.argv) > 1 else "/tmp/ueb-quitus-qr-review")
manifest = json.loads((root / "manifest.json").read_text())
headers = (
    "RÉPUBLIQUE DU CAMEROUN", "Paix – Travail – Patrie", "UNIVERSITÉ D'EBOLOWA",
    "REPUBLIC OF CAMEROON", "Peace – Work – Fatherland", "THE UNIVERSITY OF EBOLOWA",
)
count = 0
for entry in manifest:
    path = root / entry["fichier"]
    q = entry["quitus"]
    reader = PdfReader(path, strict=True)
    assert len(reader.pages) == 1, path
    expected_color = tuple(int(entry["couleur"][i:i + 2], 16) for i in (1, 3, 5))
    state = {"color": None}
    found = Counter()

    def before(operator, args, cm, tm):
        if operator == b"rg":
            state["color"] = tuple(round(float(c) * 255) for c in args)

    def visit(text, cm, tm, font, size):
        for header in headers:
            if header in text:
                assert state["color"] == expected_color, (path, header, state["color"])
                found[header] += 1

    text = reader.pages[0].extract_text(visitor_operand_before=before, visitor_text=visit)
    assert all(found[header] == 4 for header in headers), (path, found)
    assert text.count(q["numero"]) == 4, path
    assert text.count(q["identifiant"]) == 4, path
    with pdfium.PdfDocument(str(path)) as pdf:
        page = pdf[0]
        rendered = page.render(scale=300 / 72).to_pil()
        codes = zxingcpp.read_barcodes(rendered)
        assert len(codes) == 4, (path, "QR décodés", len(codes))
        for code in codes:
            assert code.text == entry["qr"], (path, "contenu QR différent")
            for field in ("numero", "identifiant", "nom", "prenom", "lieu_naissance",
                          "sexe", "nationalite", "departement", "parcours", "annee_academique"):
                assert str(q[field]) in code.text, (path, field)
            assert "12/04/2002" in code.text, path
            amount = f'{q["montant"]:,}'.replace(",", " ") + " FCFA"
            assert amount in code.text, path
            identifier = "Matricule :" if q["type_identifiant"] == "matricule" else "N° de dossier :"
            assert identifier in code.text, path
            if q["type"] == "medicaux":
                assert "Frais médicaux · Paiement unique" in code.text, path
                assert "Tranche" not in code.text, path
            else:
                tranche = "Tranches 1 et 2" if q["tranche"] == 3 else "Tranche 2"
                assert "Droits universitaires · " + tranche in code.text, path
            assert "http" not in code.text, path
            assert q["email"] not in code.text and q["nom_urgence"] not in code.text, path
            count += 1
        if entry["fichier"] in ("FSJP-droits.pdf", "FSEG-medicaux.pdf", "identite-longue.pdf"):
            rendered.save(root / (path.stem + ".png"))
        page.close()

assert len(PdfReader(root / "quitus-couleurs-qr-18-pages.pdf").pages) == 18
print(f"{len(manifest)} PDF vérifiés : couleurs bilingues et {count} QR décodés à 300 dpi, "
      "avec accents, identifiants dossier/matricule et identité longue.")
